### ISSUE TITLE
Run unit tests as part of CI builds

### ISSUE THREAD

**Issue by mccheah** [apache-spark-on-k8s/spark#110]
Currently we're relying on integration tests with Minikube to verify the correctness of our work. We should also include lower-level unit tests to harden our current behavior and prevent regressions of it.

This will be far easier once we do #109.

**Comment by kimoonkim** [comment]
+1. Do we think we should run unit tests only in kubernetes maven modules first? I find other spark module unit tests flaky and take too much time.

**Comment by mccheah** [comment]
I feel pretty comfortable with running `core` and `kubernetes`. We may make some changes to `core` down the line but it should be better than running `streaming`, etc.

**Comment by kimoonkim** [comment]
Makes sense. Maybe try them in Travis CI even before #109? If yes, I can write a PR.

**Comment by mccheah** [comment]
Sure, I think we can unit test `KubernetesClusterSchedulerBackend` pretty sanely right now. `Client` is the class that probably needs some work to make it easier to test.

**Comment by kimoonkim** [comment]
Sounds good. I'll see what I can do.

**Comment by kimoonkim** [comment]
Thinking about using the following:
```
$ build/mvn -Pmesos -Pyarn -Phadoop-2.3 -Pkubernetes  \
   -pl core,resource-managers/kubernetes/core -am clean test
```
, which includes `core` and `resource-managers/kubernetes/core`. (A few other modules they depend on will also run, but they are fast) From my Macbook:

```
[INFO] Reactor Summary:
[INFO]
[INFO] Spark Project Parent POM ........................... SUCCESS [  6.051 s]
[INFO] Spark Project Tags ................................. SUCCESS [  8.760 s]
[INFO] Spark Project Networking ........................... SUCCESS [01:00 min]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [ 15.120 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [ 14.343 s]
[INFO] Spark Project Launcher ............................. SUCCESS [ 14.200 s]
[INFO] Spark Project Core ................................. SUCCESS [21:36 min]
[INFO] Spark Project Kubernetes ........................... SUCCESS [ 14.804 s]
```

I also explored using -DwildcardSuites with package name filters. Hard to find good filters for `core` because too many spark modules use overlapping package names.

@mccheah Does this make sense? I imagine I can put this in .travis.yml and run it in parallel with the existing basic maven command.

**Comment by ash211** [comment]
This looks reasonable to enable -- we don't have many (any?) unit tests right now but starting to run them now makes sense.

@mccheah would you support adding the command @kimoonkim pasted above to the travis config?  Looks like it would run just the tests.  I'd want to make one change: hadoop 2.7 instead of 2.3, but that won't be a large change.

**Comment by kimoonkim** [comment]
@ash211 @mccheah I started writing a .travis.yml change assuming we want to do this. I'll switch to hadoop 2.7 there since it's quite trivial.

**Comment by kimoonkim** [comment]
Played with [Travis on my personal fork](https://travis-ci.org/kimoonkim/spark) a bit. Turned out the spark `core` unit tests are flaky. [Hangs and times out](https://travis-ci.org/kimoonkim/spark/jobs/202093117) too often causing build failures. I am thinking about dropping spark `core` all together or filtering even more so only a subset related with kubernetes client code runs.

**Comment by mccheah** [comment]
We're going to want to test `core` since we make changes to it - #116 comes to mind.

**Comment by kimoonkim** [comment]
@mccheah Thanks for the input. It seems most of our changes to `core` so far are made on a small set of packages, like `org.apache.spark.deploy`.

```
$ wget -qO- https://github.com/apache-spark-on-k8s/spark/pull/1.diff | grep "^+++ b/" | grep "src\/main\/scala" | sort | uniq
+++ b/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala
+++ b/core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala
+++ b/core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionServer.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesClientBuilder.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/LoggingPodStatusWatcher.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/constants.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/KubernetesRestProtocolMessages.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/CompressionUtils.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/HttpClientUtil.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesFileUtils.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestApi.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/KubernetesSparkRestServer.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/MultiServerFeignTarget.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterManager.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala
+++ b/resource-managers/kubernetes/integration-tests-spark-jobs/src/main/scala/org/apache/spark/deploy/kubernetes/integrationtest/jobs/FileExistenceTest.scala
+++ b/resource-managers/kubernetes/integration-tests-spark-jobs/src/main/scala/org/apache/spark/deploy/kubernetes/integrationtest/jobs/SparkPiWithInfiniteWait.scala
```

Maybe we can start by including only these packages, using `-DwildcardSuite` in tandem with `-pl`? And add more as needed (This needs some maintenance work, I admit)

>$ build/mvn test -Pmesos -Pyarn -Phadoop-2.7 -Pkubernetes  \
>   **-pl core,resource-managers/kubernetes/core** -am  \
> -Dtest=none **-DwildcardSuites=org.apache.spark.deploy.,org.apache.spark.scheduler.cluster.kubernetes.**

I tried this in my Travis ([an example build](https://api.travis-ci.org/jobs/202406734)). This covers [ClientSuite](https://github.com/apache-spark-on-k8s/spark/blob/k8s-support-alternate-incremental/core/src/test/scala/org/apache/spark/deploy/), [SparkSubmmitSuite](https://github.com/apache-spark-on-k8s/spark/blob/k8s-support-alternate-incremental/core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala), etc. It seems more stable than running all unit tests in `core`.

The point is to avoid flaky unit test builds, which slows down everyone. But, I am open to other alternatives. Any suggestions?

**Comment by mccheah** [comment]
What about running the full suite for now and then blacklisting the specific ones we find to be flaky as we go along? This is a more conservative stance where we try to test as much as we can and have minimal risk of omitting a test that would have caught something we introduced. But I don't know how difficult it would be to specify specific tests to omit from the Maven build; we wouldn't want to use annotations on the test classes themselves.

**Comment by kimoonkim** [comment]
We can try that. Yes, we don't want to annotate source code. It all boils down to what kind of negative filters are supported by the scalatest maven plugin. I couldn't find good options to use so far, but we can dig more. Also hunting down flaky tests can be time-consuming in general. But if we're lucky, maybe it's just one or two.

Another thing is whether we can drop JDK7 builds. With the unit tests config added, Travis will run four parallel builds (Unit tests and full builds for JDK7. Same for JDK8). I have a hunch running all of them adds to flakiness. I'd rather drop JDK7 and buy more stability and coverage instead if I can. What do you think?

**Comment by ash211** [comment]
I'd support dropping JDK7 support in the builds -- Apache Spark seems to be planning to drop it in 2.2.0 anyway: https://issues.apache.org/jira/browse/SPARK-19493

Also we are working through getting Spark to run on CircleCI (similar to Travis) in another repo: https://github.com/palantir/spark/pull/100  I'd be on the lookout for test flakes coming from those tests as we've been having trouble with them on CircleCI also.

**Comment by kimoonkim** [comment]
@ash211 Ah, that's good to know. I think dropping JDK7 will help. Thanks!

**Comment by kimoonkim** [comment]
Found [dev/run-tests](https://github.com/apache-spark-on-k8s/spark/blob/k8s-support-alternate-incremental/dev/run-tests.py) referred to by the Spark project [README.md](https://github.com/apache-spark-on-k8s/spark/blob/master/README.md). It seems to do more intelligent filtering. It also supports running only those unit tests affected by a change. Anyone knows about this? Worth looking?

**Comment by kimoonkim** [comment]
`dev/run-tests` does not meet our need. Running only affected unit tests triggers only in Amplab Jenkins env. Faking the Amplab env and then touching `core`, I found it triggering unit tests across all modules (It has this logic, touching core warrants re-testing everything).

**Comment by kimoonkim** [comment]
Found a [stackoverflow thread](http://stackoverflow.com/questions/11516038/how-to-disable-test-suite-in-scalatest) that sums up how one can disable test suites in scalatest. @mccheah The second option using a master suite with a black list seems interesting:

>Tags are one option, but you need to tag each tests accordingly, you cannot exclude complete suits that way. However, as user272735 mentioned, tagging tests is a one time effort.
>
>Another option is defining a master suite and limiting the list of nestedSuites to the suites you would like to run (see the docs). The obvious disadvantage is, that you have to maintain the master test suite, i.e., add newly created test suites to it. You could avoid this by recursing through your folders, creating a list of all test suits that can be found, and subtracting those that you put into a black list. In addition, recursing through your folders already gives you an opportunity the filter out files that match an (externally) given criterion.
>
>A third option - albeit more hacky - is to change the package to which the suites you want to black list belong. You can then use a test runner to only run suits included in certain packages. Read more in the docs, section "Specifying 'members-only' and 'wildcard' Suite paths".

I have set up [daily unit test builds](https://travis-ci.org/kimoonkim/spark/branches) on my personal Travis:

1. The build in branch `all-core-unit-tests` runs all `core` unit tests.
2. The other one in branch `unit-tests-package-filters` specifies a package filter `org.apache.spark.deploy`. i.e. It runs only a small subset of `core` unit tests in that package.

Both dropped JDK7.

Hopefully, they will give us more data points as to how often builds fail and which tests are flaky. So far, (1) failed once out of two tries.

**Comment by kimoonkim** [comment]
>What about running the full suite for now and then blacklisting the specific ones we find to be flaky as we go along? This is a more conservative stance where we try to test as much as we can and have minimal risk of omitting a test that would have caught something we introduced. But I don't know how difficult it would be to specify specific tests to omit from the Maven build; we wouldn't want to use annotations on the test classes themselves.

@mccheah Found a command line option for blacklisting tests: `-Dsuffixes`. FYI, this is the same scalatest plugin config key we used in #93. It supports full Java regex. So we can use the following for banning, say, `LocalCheckpointSuite` and `StandaloneDynamicAllocationSuite`:
```
$ build/mvn -Pmesos -Pyarn -Phadoop-2.3 -Pkubernetes  \
    -pl core,resource-managers/kubernetes/core -am clean test  \
    -Dsuffixes='^(?!org\.apache\.spark\.rdd\.LocalCheckpointSuite$|org\.apache\.spark\.deploy\.StandaloneDynamicAllocationSuite$).*'
```
BTW, the two suites are what Travis found flaky so far. I'll see it will get happier with this command line.

**Comment by mccheah** [comment]
Nice - thanks a lot for the investigation so far, let us know how this turns out!

**Comment by kimoonkim** [comment]
So far so good. Three tries all passed! I think I can send this as PR.

**Comment by kimoonkim** [comment]
With PR #132 ready for review, I got the [first flaky failure](https://travis-ci.org/kimoonkim/spark/jobs/204351605) from my daily Travis build with this change in it. (This is first out of four tries):

```
SparkSubmitSuite:
...
- includes jars passed in through --jars *** FAILED ***
  Process returned with exit code 1. See the log4j logs for more detail. (SparkSubmitSuite.scala:674)
```

Looking at the test code, the root cause might be timeout inside `runSparkSubmit`:
```
  test("includes jars passed in through --jars") {
    val unusedJar = TestUtils.createJarWithClasses(Seq.empty)
    val jar1 = TestUtils.createJarWithClasses(Seq("SparkSubmitClassA"))
    val jar2 = TestUtils.createJarWithClasses(Seq("SparkSubmitClassB"))
    val jarsString = Seq(jar1, jar2).map(j => j.toString).mkString(",")
    val args = Seq(
      "--class", JarCreationTest.getClass.getName.stripSuffix("$"),
      "--name", "testApp",
      "--master", "local-cluster[2,1,1024]",
      "--conf", "spark.ui.enabled=false",
      "--conf", "spark.master.rest.enabled=false",
      "--jars", jarsString,
      unusedJar.toString, "SparkSubmitClassA", "SparkSubmitClassB")
    runSparkSubmit(args)
  }

...

 private def runSparkSubmit(args: Seq[String]): Unit = {
    val sparkHome = sys.props.getOrElse("spark.test.home", fail("spark.test.home is not set!"))
    val sparkSubmitFile = if (Utils.isWindows) {
      new File("..\\bin\\spark-submit.cmd")
    } else {
      new File("../bin/spark-submit")
    }
    val process = Utils.executeCommand(
      Seq(sparkSubmitFile.getCanonicalPath) ++ args,
      new File(sparkHome),
      Map("SPARK_TESTING" -> "1", "SPARK_HOME" -> sparkHome))

    try {
      val exitCode = failAfter(60 seconds) { process.waitFor() }
      if (exitCode != 0) {
        fail(s"Process returned with exit code $exitCode. See the log4j logs for more detail.")
      }
    } finally {
      // Ensure we still kill the process in case it timed out
      process.destroy()
    }
  }
```

Unfortunately, this suite is not something we want to simply exclude because we actually care. i.e. This represents a case we'll have to consider fixing the flakiness.

**Comment by mccheah** [comment]
We may not be too concerned about SparkSubmitSuite - we should be integration testing the kubernetes-specific options in our own tests. `KubernetesSuite` effectively does this.

**Comment by mccheah** [comment]
But in this particular case I'm a little surprised that the test failed in the first place. This doesn't seem like something that would be expected to be flaky. Does upstream Spark have a list of tests that they know are flaky?

**Comment by kimoonkim** [comment]
I have no idea whether upstream Spark maintains a list of flaky tests. Maybe we should explore this a bit.

Personally, I am not so surprised that this test is flaky because it seems to launch a real process and relies on time out. (Strictly speaking, this is more like integration tests, not unit tests. Relying on external resources outside your control is not a good idea for unit tests) And, it only failed once out of 5 tries. Low frequency flaky tests are worst. Most people would not encounter the failure and thus did not have a chance to fix it.

**Comment by kimoonkim** [comment]
> I have no idea whether upstream Spark maintains a list of flaky tests. Maybe we should explore this a bit.

We can issue a [Jira query](https://issues.apache.org/jira/browse/SPARK-7354?jql=project%20%3D%20SPARK%20AND%20status%20in%20(Open%2C%20%22In%20Progress%22%2C%20Reopened)%20AND%20component%20%3D%20%22Spark%20Core%22%20AND%20text%20~%20%22flaky%22) that searches "flaky" in spark `core`. And guess what the first result is :-)

- SPARK-7354 Flaky test: o.a.s.deploy.**SparkSubmitSuite --jars**
- SPARK-17481 Flaky test: org.apache.spark.DistributedSuite.passing environment variables to cluster
- SPARK-18468 Flaky test: org.apache.spark.sql.hive.HiveSparkSubmitSuite.SPARK-9757 Persist Parquet relation with decimal column
- SPARK-12430 Temporary folders do not get deleted after Task completes causing problems with disk space.

SPARK-7354 was added in Dec 2015, and is still open. ~~I'll update my PR to exclude `SparkSubmitSuite`.~~

**Comment by kimoonkim** [comment]
Since the current PR #132 is merged in as is, I'll send a new PR to exclude `SparkSubmitSuite`.

**Comment by foxish** [comment]
I think we're seeing the effects of flaky tests in the CI builds on the open PRs.
